### PR TITLE
Feature/threads privacy and filtering

### PR DESCRIPTION
**Pull request by orndorffgrant**
If this PR fixes a bug, you _must_ add test cases representative of the bug.
#### What's this PR do?
1. Adds tests for the `threads` endpoints (not to be confused with the `thread` endpoints, which support a completely separate thread implementation)
2. Fixes a privacy/security issue that the `threads` endpoints had (any authenticated user could read any thread - even if they weren't included)
3. Adds a `start_date` filter to the "show thread" `threads` endpoint to allow fetching of only messages after the supplied `start_date`
4. Adds a `.eslintignore` file so that eslint will stop complaining about everything in `dist/`
#### Related JIRA tickets:
Orange-300
#### How should this be manually tested?
1. run `yarn test` and verify all tests pass
2. Via postman (or any other REST client of your choice!):
    * create three users (via auth service): user0, user1, and user2
    * as user0, create a thread between user0 and user1
      * `POST /threads { participants: ['user0', 'user1'], subject: '', topic: '', message: 'msg' }`
    * as user2, try to get messages of, view participants of, and reply to that thread
      * `GET /threads/thread/:threadId`
      * `GET /threads/thread/participants/:threadId`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
      * all of these should return 404
3. Via postman (or any other REST client of your choice!):
    * Create a thread with a couple messages
      * `POST /threads { participants: ['user0', 'user1'], subject: '', topic: '', message: 'msg' }`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
    * Get the messages of that thread using a start_date filter
      * `GET /threads/thread/:threadId?start_date=${date}`
      * verify only messages at or after the date you provided are returned
4. Run `yarn lint` and verify there are no linting errors
    * introduce a linting error in a random file
    * Run `yarn lint` and verify it finds the error you introduced

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,11 +16,11 @@
     "create_db": "cross-env NODE_ENV=test node dist/create_db.js",
     "delete_db": "cross-env NODE_ENV=test node dist/delete_db.js",
     "build": "gulp",
-    "lint": "esw -c .eslintrc *.js server config . --color --fix",
+    "lint": "esw -c .eslintrc --color --fix .",
     "lint:watch": "yarn lint -- --watch",
     "beautify": "esw -c .eslintrc *.js server config --color --fix",
     "precommit": "yarn lint && yarn test",
-    "test": "gulp && yarn create_db && cross-env NODE_ENV=test ./node_modules/.bin/mocha --ui bdd --reporter spec --colors --compilers js:babel-core/register server/tests --recursive || yarn delete_db",
+    "test": "gulp && yarn create_db && cross-env NODE_ENV=test ./node_modules/.bin/mocha --ui bdd --reporter spec --colors --compilers js:babel-core/register server/tests --recursive; yarn delete_db",
```
Comment (dbrezack) [inline package.json]: This caused an issue on my windows machine, why was this changed, I assume it has something to do with the mac terminal ?

Comment (orndorffgrant) [reply package.json]: Forgot I messed with that. The problem I had was the database wasn't being deleted when the tests passed due to the `||`. But if the semicolon doesn't work on windows, I'm not sure what to do.

maybe something like:

```
gulp && yarn create_db && cross-env NODE_ENV=test ./node_modules/.bin/mocha --ui bdd --reporter spec --colors --compilers js:babel-core/register server/tests --recursive && yarn delete_db || yarn delete_db
```

?

File: server/controllers/threads.controller.js [hunk 2]
```diff
@@ -203,14 +231,28 @@ function show(req, res, next) {
...
             });
         });
-        thread.getMessages({
+
+        // Filter by date if start_date query param is given
+        const where = {};
+        if (req.query.start_date) {
+            const startDate = new Date(req.query.start_date);
+            where.where = {
+                updatedAt: {
```
Comment (nickmccally) [inline server/controllers/threads.controller.js:240]: I would change this from `updatedAt` to `createdAt`. Few explainations, we never want to change ordering of messages regardless if we add functionality to edit messages (think how a forum allows edits) and we want the chronological order to remain the same, if you think of a messaging service like slack or facebook messenger.

File: server/tests/common.integration.js [hunk 3]
```diff
@@ -3,6 +3,6 @@ import config from '../../config/config';
 
 const version = p.version.split('.').shift();
```
Comment (nickmccally) [inline server/tests/common.integration.js:4]: Tests would not run, maybe due to this new code? `throw new Error('Config validation error: ' + error.message);`

**Review by nickmccally — CHANGES_REQUESTED** [review]
@orndorffgrant--  code looks great so far. I have requested a change for the date reference column. I also failed to run tests locally, was there anything specific that would cause this with new tests that were added?

File: server/controllers/threads.controller.js (hunk 2 above)
Comment (orndorffgrant) [reply server/controllers/threads.controller.js:240]: Sure that makes sense, I'll make this change now.

File: server/tests/common.integration.js (hunk 3 above)
Comment (orndorffgrant) [reply server/tests/common.integration.js:4]: Hmm probably a .env issue, I'll look at my env and send you what i'm using that should work
